L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,425 @@
+    const TlsServerCredentialsOptions& options);
+```
+
+Here is an example of how to configure channel credentials:
+```c++
+/* Create option and set basic security primitives. */
+TlsChannelCredentialsOptions options = TlsChannelCredentialsOptions();
+options.set_verify_server_cert(true);
+options.set_min_tls_version(TLS1_2);
+options.set_max_tls_version(TLS1_3);
+
+/* Create TLS channel credentials. */
+std::shared_ptr<ChannelCredentials> creds = TlsCredentials(options);
+```
+
+### Credential Reloading
+Credential reloading will be implemented via a `CertificateProviderInterface`.
+This provider is responsible for sourcing key material and supplying it to the
+internal stack via the `CertificateProvider`'s `SetKeyMaterials` API.  Two
+reference implementations of the `CertificateProviderInterface` are provided -
+`StaticDataCertificateProvider` and `FileWatcherCertificateProvider`. The
+`StaticDataCertificateProvider` provides certificates from raw string data. The
+`FileWatcherCertificateProvider` will periodically and asynchronously refresh
+certificate data from a file, allowing changes to certificates without
+restarting the client/server. More detail about these reference implementations
+is provided below. These implementations should cover many use cases.  If these
+providers do not support the intricacies for a specific use case, a user can
+provide their own implementations of the `CertificateProviderInterface`.
+
+```c
+/* Opaque types. */
+// A struct that stores the credential data presented to the peer in handshake
+// to show local identity. The private_key and certificate_chain must be PEM
+// encoded and the public key in the leaf certificate must correspond to the
+// given private key.
+struct IdentityKeyCertPair {
+  std::string private_key_pem;
+  std::string certificate_chain_pem;
+};
+
+
+/** ------------------------------------ Provider ------------------------------------ */
+/** Provides identity credentials and root certificates.
+*/
+class CertificateProviderInterface {
+ public:
+  virtual ~CertificateProviderInterface() = default;
+  // Provider implementations MUST provide a WatchStatusCallback that will be
+  // called by the internal stack. This will be invoked when a new certificate
+  // name is starting to be used internally, or when a certificate name is no
+  // longer being used internally. When being invoked for a new certificate,
+  // this callback should call `SetKeyMaterials` to do the initial population
+  // the certificate data in the internal stack.
+  // 
+  // For the parameters in the callback function:
+  // cert_name The name of the certificates being watched.
+  // root_being_watched If the root certificates with the specific name are being
+  // watched.
+  // identity_being-Watched If the identity certificates with the specific name
+  // are being watched.
+  virtual void WatchStatusCallback(std::string cert_name, bool root_being_watched, bool identity_being_watched) = 0;
```

Comment:
Users need to implement this can call SetKeyMaterials so that the distributor learns about the credentials.

Reply:
Where are we going to call this API? We will call this in the same place where we create the distributor.

Reply:
- Have an `Init` API that creates the distributor and calls this callback.
- The user could call this in the constructor of the derived class or in their own code.

- In Init, we're setting the watch status function to this function, not calling it.

- Conceptually this "LoadAndSetCredentialsCallback", where "setting credentials" is delegated to "SetKeyMaterials"

- What goes wrong if we do "void LoadAndSetCredentialsCallback(const std::string& name, XXX type);"?